### ISSUE TITLE
Boolean config parameters in manual thing definitions

### ISSUE THREAD

**Issue by mhilbush** [eclipse/smarthome#4452]
If you have a config parameter that is defined like this:
```
<parameter name="boolParameter" type="boolean">
	<label>Boolean Config Parameter</label>
	<description>This parameter is defined with type boolean</description>
	<default>false</default>
</parameter>
```
And, if the Thing is created manually using a .things file containing this:
```
some:thing:1 [ boolParameter="true" ]
```
instead of this
```
some:thing:2 [ boolParameter=true ]
```
Should `some:thing:1` be created even though the config parameter value is a string rather than a boolean?

Or, should the binding check the type returned by `thing.getConfiguration().get("boolParameter")` and deal with either possibility (String or Boolean)?

Currently, `some:thing:1` is created (without any warning), and it looks completely correct in PaperUI, yet the parameter value returned by `thing.getConfiguration().get("boolParameter")` is of type String.

**Comment by htreu** [comment]
I would expect the framework to check the type and return strict Boolean in this case. With special Strings "true" and "false" being converted to `true`and `false` boolean values.
There is already `org.eclipse.smarthome.config.core.ConfigUtil.normalizeType(Object, ConfigDescriptionParameter)` which is supposed to do exactly that using the `org.eclipse.smarthome.config.core.internal.normalization.BooleanNormalizer`.

Can you evaluate why this is not happening for the manual thing configuration?

**Comment by mhilbush** [comment]
Thanks. That info is helpful.  I'll see if I can track this down.

**Comment by mhilbush** [comment]
@htreu 

From what I can see (and I might be way off on this because I've never looked at this code before), the config is created without knowledge of the config parameter type as defined in the thing XML; hence it cannot be normalized to the type specified in the thing XML.

Please validate my thought process below.

The config is created here:
https://github.com/eclipse/smarthome/blob/a4dea98a1c91a36d4dbe6d5d616879988f176459/bundles/model/org.eclipse.smarthome.model.thing/src/org/eclipse/smarthome/model/thing/internal/GenericThingProvider.xtend#L414

The config values are put here; however, note that normalizeType is called with a single parameter:
https://github.com/eclipse/smarthome/blob/master/bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/Configuration.java#L152

Then here, normalizeType is called with the second parameter null.
https://github.com/eclipse/smarthome/blob/master/bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/ConfigUtil.java#L64

Which, if I understand this correctly, the config parameter value is normalized; however, it is normalized without any knowledge of the config parameter type is as specified in the thing xml.  In my example above, the value "true" is normalized to a String, because there is no knowledge that the parameter is defined in the thing XML as a Boolean.

**Comment by htreu** [comment]
Yes, I can confirm your observations. @sjka is this just not implemented yet, is type save configuration from .things file user input dependent or is this working as intended?

**Comment by sjka** [comment]
Without digging deeper into it this looks like a bug to me. If there is a config description available then the normalization should make use of it.
